src/events.rs

```diff
@@ -0,0 +1,296 @@
+use std::sync::Arc;
+
+use crate::worker;
+use futures::future::{BoxFuture, Future, FutureExt};
+use tokio::sync::{mpsc, Mutex, MutexGuard};
+use tokio::task::{self, JoinHandle};
+
+/// Event represents all the different events that may happen on a running
+/// supervision tree.
+#[derive(Debug, Clone)]
+pub enum Event {
+    SupervisorStarted(NodeData),
+    SupervisorStartFailed(NodeData, Arc<anyhow::Error>),
+    SupervisorTerminated(NodeData),
+    SupervisorTerminationFailed(NodeData, Arc<anyhow::Error>),
+    WorkerStarted(NodeData),
+    WorkerStartFailed(NodeData, Arc<worker::StartError>),
+    WorkerTerminated(NodeData),
+    WorkerTerminationFailed(NodeData, Arc<worker::TerminationError>),
+}
+
+/// NodeData holds details about the producer of the event (supervisor or worker)
+#[derive(Debug, Clone, PartialEq)]
+pub struct NodeData {
+    pub runtime_name: String,
+}
+
+/// NotifyFn is used by the supervision API to send events to an interested
+/// listener.
+type NotifyFn = Box<dyn Fn(Event) -> BoxFuture<'static, ()> + Send>;
+
+/// EventNotifier is used by the internal supervision API to send events about a
+/// running supervision tree
+#[derive(Clone)]
+pub struct EventNotifier(Arc<Mutex<NotifyFn>>);
+
+impl EventNotifier {
+    pub fn new<F, O>(notify0: F) -> Self
+    where
+        F: Fn(Event) -> O + Send + 'static,
```

Comment:
I remove the `Send` constraint and I get this error

```
error[E0277]: `F` cannot be sent between threads safely
  --> src/events.rs:47:32
   |
43 |           let notify = move |ev| {
   |  ______________________-
44 | |             let fut = notify0(ev);
45 | |             fut.boxed()
46 | |         };
   | |_________- within this `[closure@src/events.rs:43:22: 46:10 notify0:F]`
47 |           EventNotifier(Arc::new(Box::new(notify)))
   |                                  ^^^^^^^^^^^^^^^^ `F` cannot be sent between threads safely
   |
   = help: within `[closure@src/events.rs:43:22: 46:10 notify0:F]`, the trait `std::marker::Send` is not implemented for `F`
   = note: required because it appears within the type `[closure@src/events.rs:43:22: 46:10 notify0:F]`
   = note: required for the cast to the object type `dyn std::ops::Fn(events::Event) -> std::pin::Pin<std::boxed::Box<(dyn futures::Future<Output = ()> + std::marker::Send + 'st
atic)>> + std::marker::Send`
help: consider further restricting this bound
   |
40 |         F: Fn(Event) -> O + 'static + std::marker::Send,
   |                                     ^^^^^^^^^^^^^^^^^^^
```

Reply:
Agh, nevermind, it was a conflict on the type declaration